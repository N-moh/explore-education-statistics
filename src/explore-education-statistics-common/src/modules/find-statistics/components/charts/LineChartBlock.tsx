import {
  ChartDataB,
  ChartDefinition,
  ChartProps,
  createDataForAxis,
  getKeysForChart,
  mapNameToNameLabel,
  populateDefaultChartProps,
} from '@common/modules/find-statistics/components/charts/ChartFunctions';

import React, { Component } from 'react';
import {
  AxisDomain,
  CartesianGrid,
  Legend,
  Line,
  LineChart,
  ResponsiveContainer,
  Symbols,
  Tooltip,
  TooltipProps,
  XAxis,
  YAxis,
} from 'recharts';
import LoadingSpinner from '@common/components/LoadingSpinner';

const CustomToolTip = ({ active, payload, label }: TooltipProps) => {
  if (active) {
    return (
      <div className="graph-tooltip">
        <p>{label}</p>
        {payload &&
          payload
            .sort((a, b) => {
              if (typeof b.value === 'number' && typeof a.value === 'number') {
                return b.value - a.value;
              }

              return 0;
            })
            .map((_, index) => {
              return (
                // eslint-disable-next-line react/no-array-index-key
                <p key={index}>
                  {`${payload[index].name} : ${payload[index].value}`}
                </p>
              );
            })}
      </div>
    );
  }

  return null;
};

export default class LineChartBlock extends Component<ChartProps> {
  public static definition: ChartDefinition = {
    type: 'line',
    name: 'Line',

    data: [
      {
        type: 'line',
        title: 'Line',
        entryCount: 'multiple',
        targetAxis: 'xaxis',
      },
    ],

    axes: [
      {
        id: 'xaxis',
        title: 'X Axis',
        type: 'major',
        defaultDataType: 'timePeriods',
      },
      {
        id: 'yaxis',
        title: 'Y Axis',
        type: 'minor',
      },
    ],
  };

  public render() {
    const { data, meta, height, axes, labels } = this.props;

    const yAxisDomain: [AxisDomain, AxisDomain] = [-10, 10];

    if (axes.major && data) {
      const chartData: ChartDataB[] = createDataForAxis(
        axes.major,
        data.result,
        meta,
      ).map(mapNameToNameLabel(labels, meta.timePeriods, meta.locations));

      const keysForChart = getKeysForChart(chartData);

      return (
        <ResponsiveContainer width={900} height={height || 300}>
          <LineChart
            data={chartData}
            margin={{ top: 5, right: 30, left: 20, bottom: 25 }}
          >
            <Tooltip content={CustomToolTip} />
            <Legend verticalAlign="top" height={36} />
            <CartesianGrid
              strokeDasharray="3 3"
              horizontal={axes.minor.showGrid !== false}
              vertical={axes.major.showGrid !== false}
            />

            {axes.major && axes.major.visible && (
              <XAxis
                dataKey="name"
                label={{
                  offset: 5,
                  position: 'bottom',
                  value: '',
                }}
                scale="auto"
                interval={
                  axes.minor && !axes.minor.visible
                    ? 'preserveStartEnd'
                    : undefined
                }
                padding={{ left: 20, right: 20 }}
                tickMargin={10}
              />
            )}

            {axes.minor && axes.minor.visible && (
              <YAxis
                label={{
                  angle: -90,
                  offset: 0,
                  position: 'left',
                  value: '',
                }}
                scale="auto"
                domain={yAxisDomain}
                dataKey="value"
              />
            )}

            {keysForChart.map(name => (
              <Line
                key={name}
                {...populateDefaultChartProps(name, labels[name])}
                type="linear"
                legendType={labels[name] && labels[name].symbol}
                dot={
                  labels[name] &&
                  labels[name].symbol &&
                  (props => (
                    <Symbols
                      {...props}
                      type={labels[name] && labels[name].symbol}
                    />
                  ))
                }
                strokeWidth="2"
              />
            ))}
          </LineChart>
        </ResponsiveContainer>
      );
    }

    return <LoadingSpinner />;
  }
}
