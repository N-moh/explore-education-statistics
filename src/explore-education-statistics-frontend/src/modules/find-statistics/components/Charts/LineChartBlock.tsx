import React, { Component } from 'react';
import {
  AxisDomain,
  CartesianGrid,
  Legend,
  Line,
  LineChart,
  Symbols,
  Tooltip,
  XAxis,
  YAxis,
} from 'recharts';
import { Axis } from '../../../../services/publicationService';
import { CharacteristicsData } from '../../../../services/tableBuilderService';
import { colours, parseCondensedYearRange, symbols } from './Charts';

interface LineChartBlockProps {
  chartDataKeys: string[];
  characteristicsData: CharacteristicsData;
  height?: number;
  yAxis: Axis;
  xAxis: Axis;
  labels: { [name: string]: string };
}

const CustomToolTip = (props: any) => {
  if (props.active) {
    const { payload, label } = props;

    return (
      <div className="graph-tooltip">
        <p>{label}</p>
        {payload
          .sort((a: any, b: any) => {
            return b.value - a.value;
          })
          .map((_: any, index: number) => (
            <p key={index}>{`${payload[index].name} : ${
              payload[index].value
            }`}</p>
          ))}
      </div>
    );
  }
};

export class LineChartBlock extends Component<LineChartBlockProps> {
  public render() {
    const {
      characteristicsData,
      chartDataKeys,
      xAxis,
      yAxis,
      labels,
    } = this.props;

    const chartData = characteristicsData.result.map(result => {
      return chartDataKeys.reduce(
        (v: any, attributeName) => {
          if (result.attributes[attributeName]) {
            v[attributeName] = result.attributes[attributeName];
          }
          return v;
        },
        { name: parseCondensedYearRange(`${result.year}`) },
      );
    });

    let yAxisDomain: any;

    if (yAxis.min !== undefined && yAxis.max !== undefined) {
      yAxisDomain = [yAxis.min, yAxis.max];
    }

    return (
      <LineChart
        width={900}
        height={this.props.height || 300}
        data={chartData}
        margin={{ top: 5, right: 30, left: 20, bottom: 25 }}
      >
        <Tooltip content={CustomToolTip} />
        {chartDataKeys.length > 1 ? (
          <Legend verticalAlign="top" height={36} />
        ) : (
          ''
        )}
        <CartesianGrid strokeDasharray="3 3" />
        <XAxis
          dataKey="name"
          label={{
            offset: 5,
            position: 'bottom',
            value: xAxis.title,
          }}
          padding={{ left: 20, right: 20 }}
          tickMargin={10}
        />
        <YAxis
          label={{
            angle: -90,
            offset: 0,
            position: 'left',
            value: yAxis.title,
          }}
          scale="auto"
          unit="%"
          domain={yAxisDomain}
        />
        {chartDataKeys.map((dataKey: any, index: any) => (
          <Line
            key={index}
            name={labels[dataKey]}
            type="linear"
            dataKey={dataKey}
            stroke={colours[index]}
            fill={colours[index]}
            strokeWidth="5"
            unit="%"
            legendType={symbols[index]}
            activeDot={{ r: 3 }}
            dot={props => <Symbols {...props} type={symbols[index]} />}
          />
        ))}
      </LineChart>
    );
  }
}
