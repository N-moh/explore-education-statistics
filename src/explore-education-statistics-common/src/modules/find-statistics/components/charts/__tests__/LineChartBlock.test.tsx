import React from 'react';

import { render } from 'react-testing-library';
import LineChartBlock from '../LineChartBlock';

import testData from './__data__/testBlockData';

jest.mock('recharts/lib/util/LogUtils');

describe('LineChartBlock', () => {
  test('renders with correct output', () => {
    const { container } = render(
      <LineChartBlock {...testData.AbstractChartProps} />,
    );

    expect(
      container.querySelector('.xAxis text.recharts-label tspan'),
    ).toHaveTextContent('test x axis');
    expect(
      container.querySelector('.yAxis text.recharts-label tspan'),
    ).toHaveTextContent('test y axis');

    expect(
      container.querySelector(
        '.xAxis .recharts-cartesian-axis-tick:nth-child(1) text tspan',
      ),
    ).toHaveTextContent('2014/15');
    expect(
      container.querySelector(
        '.xAxis .recharts-cartesian-axis-tick:nth-child(2) text tspan',
      ),
    ).toHaveTextContent('2015/16');

    expect(
      Array.from(container.querySelectorAll('.recharts-line')).length,
    ).toBe(3);

    expect(container).toMatchSnapshot();
  });
});
