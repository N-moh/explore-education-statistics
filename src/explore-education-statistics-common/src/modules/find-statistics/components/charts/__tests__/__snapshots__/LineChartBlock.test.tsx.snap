// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LineChartBlock renders basic chart correctly 1`] = `
<div>
  <div
    class="recharts-responsive-container"
    style="width: 900px; height: 300px;"
  >
    <div
      class="recharts-wrapper"
      style="position: relative; cursor: default; width: 900px; height: 300px;"
    >
      <svg
        class="recharts-surface"
        height="300"
        version="1.1"
        viewBox="0 0 900 300"
        width="900"
      >
        <defs>
          <clippath
            id="recharts5-clip"
          >
            <rect
              height="260"
              width="810"
              x="90"
              y="10"
            />
          </clippath>
        </defs>
        <g
          class="recharts-cartesian-grid"
        >
          <g
            class="recharts-cartesian-grid-horizontal"
          >
            <line
              fill="none"
              height="260"
              stroke="#ccc"
              stroke-dasharray="3 3"
              width="810"
              x="90"
              x1="90"
              x2="900"
              y="10"
              y1="270"
              y2="270"
            />
            <line
              fill="none"
              height="260"
              stroke="#ccc"
              stroke-dasharray="3 3"
              width="810"
              x="90"
              x1="90"
              x2="900"
              y="10"
              y1="199.13776015857286"
              y2="199.13776015857286"
            />
            <line
              fill="none"
              height="260"
              stroke="#ccc"
              stroke-dasharray="3 3"
              width="810"
              x="90"
              x1="90"
              x2="900"
              y="10"
              y1="128.2755203171457"
              y2="128.2755203171457"
            />
            <line
              fill="none"
              height="260"
              stroke="#ccc"
              stroke-dasharray="3 3"
              width="810"
              x="90"
              x1="90"
              x2="900"
              y="10"
              y1="10"
              y2="10"
            />
          </g>
          <g
            class="recharts-cartesian-grid-vertical"
          >
            <line
              fill="none"
              height="260"
              stroke="#ccc"
              stroke-dasharray="3 3"
              width="810"
              x="90"
              x1="110"
              x2="110"
              y="10"
              y1="10"
              y2="270"
            />
            <line
              fill="none"
              height="260"
              stroke="#ccc"
              stroke-dasharray="3 3"
              width="810"
              x="90"
              x1="880"
              x2="880"
              y="10"
              y1="10"
              y2="270"
            />
            <line
              fill="none"
              height="260"
              stroke="#ccc"
              stroke-dasharray="3 3"
              width="810"
              x="90"
              x1="90"
              x2="90"
              y="10"
              y1="10"
              y2="270"
            />
            <line
              fill="none"
              height="260"
              stroke="#ccc"
              stroke-dasharray="3 3"
              width="810"
              x="90"
              x1="900"
              x2="900"
              y="10"
              y1="10"
              y2="270"
            />
          </g>
        </g>
        <g
          class="recharts-layer recharts-cartesian-axis recharts-yAxis yAxis"
        >
          <line
            class="recharts-cartesian-axis-line"
            fill="none"
            height="260"
            stroke="#666"
            width="60"
            x="30"
            x1="90"
            x2="90"
            y="10"
            y1="10"
            y2="270"
          />
          <g
            class="recharts-cartesian-axis-ticks"
          >
            <g
              class="recharts-layer recharts-cartesian-axis-tick"
            >
              <line
                class="recharts-cartesian-axis-tick-line"
                fill="none"
                height="260"
                stroke="#666"
                width="60"
                x="30"
                x1="84"
                x2="90"
                y="10"
                y1="270"
                y2="270"
              />
              <text
                class="recharts-text recharts-cartesian-axis-tick-value"
                fill="#666"
                height="260"
                stroke="none"
                text-anchor="end"
                width="60"
                x="82"
                y="270"
              >
                <tspan
                  dy="0.355em"
                  x="82"
                >
                  -3
                </tspan>
              </text>
            </g>
            <g
              class="recharts-layer recharts-cartesian-axis-tick"
            >
              <line
                class="recharts-cartesian-axis-tick-line"
                fill="none"
                height="260"
                stroke="#666"
                width="60"
                x="30"
                x1="84"
                x2="90"
                y="10"
                y1="199.13776015857286"
                y2="199.13776015857286"
              />
              <text
                class="recharts-text recharts-cartesian-axis-tick-value"
                fill="#666"
                height="260"
                stroke="none"
                text-anchor="end"
                width="60"
                x="82"
                y="199.13776015857286"
              >
                <tspan
                  dy="0.355em"
                  x="82"
                >
                  547
                </tspan>
              </text>
            </g>
            <g
              class="recharts-layer recharts-cartesian-axis-tick"
            >
              <line
                class="recharts-cartesian-axis-tick-line"
                fill="none"
                height="260"
                stroke="#666"
                width="60"
                x="30"
                x1="84"
                x2="90"
                y="10"
                y1="128.2755203171457"
                y2="128.2755203171457"
              />
              <text
                class="recharts-text recharts-cartesian-axis-tick-value"
                fill="#666"
                height="260"
                stroke="none"
                text-anchor="end"
                width="60"
                x="82"
                y="128.2755203171457"
              >
                <tspan
                  dy="0.355em"
                  x="82"
                >
                  1097
                </tspan>
              </text>
            </g>
            <g
              class="recharts-layer recharts-cartesian-axis-tick"
            >
              <line
                class="recharts-cartesian-axis-tick-line"
                fill="none"
                height="260"
                stroke="#666"
                width="60"
                x="30"
                x1="84"
                x2="90"
                y="10"
                y1="10"
                y2="10"
              />
              <text
                class="recharts-text recharts-cartesian-axis-tick-value"
                fill="#666"
                height="260"
                stroke="none"
                text-anchor="end"
                width="60"
                x="82"
                y="10"
              >
                <tspan
                  dy="0.355em"
                  x="82"
                >
                  2015
                </tspan>
              </text>
            </g>
          </g>
        </g>
        <g
          class="recharts-layer recharts-cartesian-axis recharts-xAxis xAxis"
        >
          <line
            class="recharts-cartesian-axis-line"
            fill="none"
            height="30"
            stroke="#666"
            width="810"
            x="90"
            x1="90"
            x2="900"
            y="270"
            y1="270"
            y2="270"
          />
          <g
            class="recharts-cartesian-axis-ticks"
          >
            <g
              class="recharts-layer recharts-cartesian-axis-tick"
            >
              <line
                class="recharts-cartesian-axis-tick-line"
                fill="none"
                height="30"
                stroke="#666"
                width="810"
                x="90"
                x1="110"
                x2="110"
                y="270"
                y1="276"
                y2="270"
              />
              <text
                class="recharts-text recharts-cartesian-axis-tick-value"
                fill="#666"
                height="30"
                stroke="none"
                text-anchor="middle"
                width="810"
                x="110"
                y="286"
              >
                <tspan
                  dy="0.71em"
                  x="110"
                >
                  2014/15
                </tspan>
              </text>
            </g>
            <g
              class="recharts-layer recharts-cartesian-axis-tick"
            >
              <line
                class="recharts-cartesian-axis-tick-line"
                fill="none"
                height="30"
                stroke="#666"
                width="810"
                x="90"
                x1="880"
                x2="880"
                y="270"
                y1="276"
                y2="270"
              />
              <text
                class="recharts-text recharts-cartesian-axis-tick-value"
                fill="#666"
                height="30"
                stroke="none"
                text-anchor="middle"
                width="810"
                x="880"
                y="286"
              >
                <tspan
                  dy="0.71em"
                  x="880"
                >
                  2015/16
                </tspan>
              </text>
            </g>
          </g>
        </g>
        <g
          class="recharts-layer recharts-line"
        >
          <path
            class="recharts-curve recharts-line-curve"
            fill="none"
            height="260"
            stroke="#3182bd"
            stroke-width="2"
            width="810"
          />
          <g
            class="recharts-layer recharts-line-dots"
          />
        </g>
        <g
          class="recharts-layer recharts-line"
        >
          <path
            class="recharts-curve recharts-line-curve"
            d="M110,269.22695738354804L880,270"
            fill="none"
            height="260"
            stroke="#ff0000"
            stroke-width="2"
            width="810"
          />
          <g
            class="recharts-layer recharts-line-dots"
          />
        </g>
        <g
          class="recharts-layer recharts-line"
        >
          <path
            class="recharts-curve recharts-line-curve"
            d="M110,268.3250743310208L880,269.09811694747276"
            fill="none"
            height="260"
            stroke="#00ff00"
            stroke-width="2"
            width="810"
          />
          <g
            class="recharts-layer recharts-line-dots"
          />
        </g>
        <g
          class="recharts-layer recharts-line"
        >
          <path
            class="recharts-curve recharts-line-curve"
            d="M110,268.9692765113974L880,269.4846382556987"
            fill="none"
            height="260"
            stroke="#0000ff"
            stroke-width="2"
            width="810"
          />
          <g
            class="recharts-layer recharts-line-dots"
          />
        </g>
      </svg>
      <div
        class="recharts-legend-wrapper"
        style="position: absolute; width: 870px; height: 50px; left: 30px; top: 10px;"
      >
        <ul
          class="recharts-default-legend"
          style="padding: 0px; margin: 0px; text-align: center;"
        >
          <li
            class="recharts-legend-item legend-item-0"
            style="display: inline-block; margin-right: 10px;"
          >
            <svg
              class="recharts-surface"
              height="14"
              style="display: inline-block; vertical-align: middle; margin-right: 4px;"
              version="1.1"
              viewBox="0 0 32 32"
              width="14"
            >
              <path
                class="recharts-legend-icon"
                d="M0,16h10.666666666666666
            A5.333333333333333,5.333333333333333,0,1,1,21.333333333333332,16
            H32M21.333333333333332,16
            A5.333333333333333,5.333333333333333,0,1,1,10.666666666666666,16"
                fill="none"
                stroke="#3182bd"
                stroke-width="4"
              />
            </svg>
            <span
              class="recharts-legend-item-text"
            >
              __name
            </span>
          </li>
          <li
            class="recharts-legend-item legend-item-1"
            style="display: inline-block; margin-right: 10px;"
          >
            <svg
              class="recharts-surface"
              height="14"
              style="display: inline-block; vertical-align: middle; margin-right: 4px;"
              version="1.1"
              viewBox="0 0 32 32"
              width="14"
            >
              <path
                class="recharts-legend-icon"
                d="M0,16h10.666666666666666
            A5.333333333333333,5.333333333333333,0,1,1,21.333333333333332,16
            H32M21.333333333333332,16
            A5.333333333333333,5.333333333333333,0,1,1,10.666666666666666,16"
                fill="none"
                stroke="#ff0000"
                stroke-width="4"
              />
            </svg>
            <span
              class="recharts-legend-item-text"
            >
              Unauthorised absence rate
            </span>
          </li>
          <li
            class="recharts-legend-item legend-item-2"
            style="display: inline-block; margin-right: 10px;"
          >
            <svg
              class="recharts-surface"
              height="14"
              style="display: inline-block; vertical-align: middle; margin-right: 4px;"
              version="1.1"
              viewBox="0 0 32 32"
              width="14"
            >
              <path
                class="recharts-legend-icon"
                d="M0,16h10.666666666666666
            A5.333333333333333,5.333333333333333,0,1,1,21.333333333333332,16
            H32M21.333333333333332,16
            A5.333333333333333,5.333333333333333,0,1,1,10.666666666666666,16"
                fill="none"
                stroke="#00ff00"
                stroke-width="4"
              />
            </svg>
            <span
              class="recharts-legend-item-text"
            >
              Overall absence rate
            </span>
          </li>
          <li
            class="recharts-legend-item legend-item-3"
            style="display: inline-block; margin-right: 10px;"
          >
            <svg
              class="recharts-surface"
              height="14"
              style="display: inline-block; vertical-align: middle; margin-right: 4px;"
              version="1.1"
              viewBox="0 0 32 32"
              width="14"
            >
              <path
                class="recharts-legend-icon"
                d="M0,16h10.666666666666666
            A5.333333333333333,5.333333333333333,0,1,1,21.333333333333332,16
            H32M21.333333333333332,16
            A5.333333333333333,5.333333333333333,0,1,1,10.666666666666666,16"
                fill="none"
                stroke="#0000ff"
                stroke-width="4"
              />
            </svg>
            <span
              class="recharts-legend-item-text"
            >
              Authorised absence rate
            </span>
          </li>
        </ul>
      </div>
      <div
        class="recharts-tooltip-wrapper"
        style="pointer-events: none; visibility: hidden; position: absolute; top: 0px; transform: translate(undefinedpx, undefinedpx);"
      />
    </div>
    <div
      style="position: absolute; width: 0px; height: 0px; visibility: hidden; display: none;"
    />
  </div>
</div>
`;
